File: Leetcode/Medium/Insert-Delete-GetRandom-O(1)/solution.py
import random
class RandomizedSet:
    def __init__(self):
        self.vals = []
        self.vals_map = {}
    def insert(self,val):
        if val in self.vals:
            return False
        self.vals.append(val)
        self.vals_map[val] = len(self.vals-1)
        return True
    def remove(self,val):
        if val not in self.vals:
            return False
        index,lastElement = self.vals_map[val],self.vals[-1]
        self.vals[index] = lastElement
        self.vals_map[lastElement] = index
        del self.vals_map[val]
        self.vals.pop()
        return True
    def getRandom(self):
        return self.vals[random.randint(0,len(self.vals)-1)]

File: Leetcode/Medium/Insert-Delete-GetRandom-O(1)/solution.js
class RandomizedSet {
    constructor() {
        this.vals = [];
        this.valsMap = new Map();
    }
    insert(val) {
        if (this.valsMap.has(val)) return false;
        this.vals.push(val);
        this.valsMap.set(val, this.vals.length - 1);
        return true;
    }
    remove(val) {
        if (!this.valsMap.has(val)) return false;
        let [index, lastElement] = [this.valsMap.get(val), this.vals[this.vals.length - 1]];
        this.vals[index] = lastElement;
        this.valsMap.set(lastElement, index);
        this.valsMap.delete(val);
        this.vals.pop();
        return true;
    }
    getRandom() {
        let index = Math.floor(Math.random() * this.vals.length);
        return this.vals[index];
    }
};